src/MySqlConnector/MySqlClient/Results/ResultSet.cs

```diff
@@ -0,0 +1,373 @@
+					ColumnDefinitions = new ColumnDefinitionPayload[columnCount];
+					m_dataOffsets = new int[columnCount];
+					m_dataLengths = new int[columnCount];
+
+					for (var column = 0; column < ColumnDefinitions.Length; column++)
+					{
+						payload = await Session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+						ColumnDefinitions[column] = ColumnDefinitionPayload.Create(payload);
+					}
+
+					payload = await Session.ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+					EofPayload.Create(payload);
+
+					LastInsertId = -1;
+					State = ResultSetState.ReadResultSetHeader;
+					break;
+				}
+			}
+			BufferState = State;
+			//await BufferEntireAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+			return this;
+		}
+
+		public async Task BufferEntireAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			while (BufferState == ResultSetState.ReadingRows || BufferState == ResultSetState.ReadResultSetHeader)
+				await BufferReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+		}
+
+		public async Task ReadEntireAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			while (State == ResultSetState.ReadingRows || State == ResultSetState.ReadResultSetHeader)
+				await ReadAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+		}
+
+		public bool Read()
+		{
+			return ReadAsync(IOBehavior.Synchronous, CancellationToken.None).GetAwaiter().GetResult();
+		}
+
+		public Task<bool> ReadAsync(CancellationToken cancellationToken) =>
+			ReadAsync(Command.Connection.AsyncIOBehavior, cancellationToken);
+
+		public async Task<bool> ReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_row?.ClearData();
+			m_row = m_readBuffer.Count > 0
+				? m_readBuffer.Dequeue()
+				: await DoReadAsync(ioBehavior, m_row, cancellationToken).ConfigureAwait(false);
+
+			if (m_row == null)
+			{
+				State = BufferState;
+				return false;
+			}
+			State = ResultSetState.ReadingRows;
+			return true;
+		}
+
+		public async Task<Row> BufferReadAsync(IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_rowBuffered?.BufferData();
+			m_rowBuffered = await DoReadAsync(ioBehavior, null, cancellationToken).ConfigureAwait(false);
+			m_readBuffer.Enqueue(m_rowBuffered);
+			return m_rowBuffered;
+		}
+
+		private async Task<Row> DoReadAsync(IOBehavior ioBehavior, Row row, CancellationToken cancellationToken)
+		{
+			// if we've already read past the end of this resultset, Read returns false
+			if (BufferState == ResultSetState.HasMoreData || BufferState == ResultSetState.NoMoreData || BufferState == ResultSetState.None)
+				return null;
+
+			var payloadTask = Session.ReceiveReplyAsync(ioBehavior, cancellationToken);
+			if (payloadTask.IsCompletedSuccessfully)
+				row = ReadAsyncRemainder(row, payloadTask.Result);
+			row = await ReadAsyncAwaited(row, payloadTask.AsTask()).ConfigureAwait(false);
```

Comment:
This looks like a bug to assign `row` twice if the condition on line 131 is true.